IQSS/dataverse
# 9063 - add session API auth mechanism with feature flag

**What this PR does / why we need it**:

As stated in #9063, for the development of the new Dataverse Single Page Application (SPA), the SPA needs to be able to authenticate against Dataverse Native API.

The final authentication mechanism for the SPA is yet to be developed. Meanwhile, **only for development purposes**, this mechanism allows to authenticate given a session (Defined by JSESSIONID cookie). Due to the **CSRF risks** of using the JSESSIONID cookie for API authentication, **this functionality should not be enabled for production purposes**.

In order to keep this functionality disabled by default and following [the docs](https://guides.dataverse.org/en/5.12.1/developers/configuration.html#adding-a-jvm-setting), a new JVMSetting has been created, with a boolean value of 'false' by default.

Being aware of @poikilotherm's work in progress of #9230, where he developed a generic mechanism for feature flags, the naming of the JVMSetting created in this PR follows the same structure he uses in his PR, considering a later refactor that will include this JVMSetting into the generic mechanism.

Edit (01/17):  After discussion with @pdurbin and @poikilotherm, we have decided to include the generic mechanism for feature flags that Oliver has worked on and its associated documentation in this PR.

Edit (01/18): New issue created for the new generic mechanism for feature flags, #9297.

TODOs.

- [x] Add feature flag for session API auth
- [x] Add generic feature flag mechanism
- [x] Add documentation for feature flag for session API auth

**Which issue(s) this PR closes**:

Closes #9063, #9297

**Special notes for your reviewer**:

I have considered to add test coverage to the AbstractApiBean changes, but I find it difficult to test the changes in isolation. The future refactor towards a filter-based module should make this easier.

**Suggestions on how to test this**:

Follow the next steps:

1. Enable the JVMSetting in _microprofile-config.properties_ file: `dataverse.feature.api-session-auth=true`
2. Build and deploy Dataverse locally.
3. Navigate to Dataverse Login Page.
4. Without logging in, obtain the current JSESSIONID cookie value from the browser (it corresponds to a Guest User).
5. Using the obtained cookie value, perform a curl command against a secured Dataverse Native API endpoint. For example: `curl --cookie "JSESSIONID=<cookie_value>" -X GET http://localhost:8080/api/datasets/2`
6. A message indicating that the guest user is not authorized should appear.
7. Again on the Dataverse Login Page, log in with valid user credentials. The JSESSIONID cookie value should be updated.
8. Repeat the same curl test but with the new cookie value.
9. The requested data should be successfully retrieved.

An additional test is to repeat the above steps but keeping `dataverse.feature.api-session-auth=false`. Whether logged in or not, requests must fail by recognizing a guest user (Default Dataverse behavior).

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No

**Is there a release notes update needed for this change?**:

Not sure

**Additional documentation**:

No
